text/0000-cargo-workspace-deduplicate.md

```diff
@@ -0,0 +1,701 @@
+package will be, in descending order:
+
+* First, if an explicit value is listed, that's used. For example `version =
+  "1.2.3"`.
+* If a workspace member is referenced, that member's value is then used. for
+  example `authors = { workspace = "other-crate" }`
+* If the `[workspace]` section defines the key, then it is automatically filled
+  in for each `[package]`.
+
+### New dependency directives
+
+Dependencies in the `[dependencies]`, `[dev-dependencies]`,
+`[build-dependencies]`, and `[target."...".dependencies]` sections will support
+the ability to reference the `[workspace.dependencies]` definition of
+dependencies. This is done with a new `workspace` key in the dependency
+directive. An example of this looks like:
+
+```toml
+[dependencies]
+log = { workspace = true }
+```
+
+The `workspace` key cannot be defined with other keys that configure the source
+of the dependency. This means you cannot define `workspace` with keys like
+`version`, `registry`, `registry-index`, `path`, `git`, `branch`, `tag`, `rev`,
+or `package`. The `workspace` key can be combined with other keys, however:
+
+* `optional` - this introduces an optional dependency as usual, as well as a
+  feature named after the key (left hand side) of the dependency directive).
+  Note that the `[workspace.dependencies]` table is not allowed to specify
+  `optional`.
+
+* `features` - this indicates, as usual, that extra features are being enabled
+  over the already-enabled features in the directive found in
+  `[workspace.dependencies]`. The result set of enabled features is the union of
+  the features specified inline with the features specified in the directive in
+  the workspace table.
+
+* `default-features` - specifying this key is a little more subtle than with
+  normal dependencies. If this key is not specified in the `[dependencies]`
+  section of a package then it does not default to as if it were `true` (like
+  omitting it typically does). Instead the value, including the implicit
+  default, is used from the `[workspace.dependencies]` table. Otherwise if the
+  value is present in a package `[dependencies]` table it boolean or's with the
+  value, including the implicit default, in `[workspace.dependencies]`. The
+  logic here for resolving `default-features` looks like:
+
+| Package directive | Workspace Directive | resolved value |
+|-------------------|---------------------|----------------|
+| *not present*     | *not present*       | `true`         |
+| *not present*     | `true`              | `true`         |
+| *not present*     | `false`             | `false`        |
+| `true`            | *not present*       | `true`         |
+| `true`            | `true`              | `true`         |
+| `true`            | `false`             | `true`         |
+| `false`           | *not present*       | `true`         |
+| `false`           | `true`              | `true`         |
+| `false`           | `false`             | `false`        |
```

Comment:
This seems incredibly confusing, and needs both explanation and motivation. Why does false with not present give true, for instance?

Reply:
Yeah I agree it's subtle, but I wasn't sure how best to define this. I tried to explain what the table would be in the above paragraph (and then show the below paragraph). My thinking was that if you do:

```toml
[workspace.dependencies]
log = { version = "0.1", default-features = false }

[dependencies]
log = { workspace = true }
```

*technically* the `workspace = true` directive says `default-features = true`.  You almost for sure don't want this though. However if you write this:

```toml
[workspace.dependencies]
log = { version = "0.1" }

[dependencies]
log = { workspace = true, default-features = false }
```

It's sort of unclear to me what the desired end state would be. The `[workspace.dependencies]` entry technically implies `default-features = true` like before.

Honestly I think no matter what we do here it's surprising behavior to someone:

* Always assume implicit `default-features = true` - this makes the first case a little too surprising in my opinion
* Only assume `default-features = true` for non-`workspace=true` deps - this runs into the confusion that you've encountered.
* Only assume `default-features = true` in the `[workspace.dependencies]` - this is what I've currently proposed here
* Never assume `default-features = true` for either `[workspace.dependencies]` or `workspace = true` - I think this goes too much against established convention to be too worthwhile.

The only other alternative I can think of is to disallow `default-features = true` within `workspace = true` directives, and assume it take the value of `[workspace.dependencies]`. That feels the same as the third option (what's proposed here), though? I'm curious what you think of this?

Reply:
What about:

* if `default-features` is specified at both the workspace and the package level, the package-level is used
* if `default-features` is only specified in either place, use that specified value
* otherwise, default to `true`

This would be the same as the current version except for the second-to-last entry: Specifying `false` on the workspace level and not specifying `default-features` on the package level would result in the default features being turned off. To be honest I still don't see how the text above the table describes the table. It sounds like what I just described another way was actually intended...

Reply:
A more succinct way of putting what this proposal currently says is:

* `[workspace.dependencies]` table has two values for `default-features`, `false` or `true`. If it's ommitted that means `true`. 
* `[dependencies]` using `workspace = true` has three values: `false`, `true`, and blank.
* The final value for `default-features` is the boolean-or ("union") of this ternary and binary value. The `false || *blank*` part is `false`.

Note that for all other dependencies in Cargo it's interpreted with the first bullet already. There's two values for `default-features` and the default is `default-features = true` if nothing is specified.

@jplatte definitely seems like an alternative! It feels pretty different though from how features work elsewhere in Cargo. Everything with features is more-or-less "let's union everything and see what we get", whereas this logic would have precedences (if both are set the package is preferred) as well as non-unioning logic (if set in one place you disregard the other place instead of trying to account for it). 

I don't really know of a reason to specifically allow `default-features` in `workspace = true` dependency directives though. It might be best to simply say that `default-features` is disallowed and its value is always inherited from `[workspace.dependencies]`. That way if the workspace disables default features and you want to enable them you specify `features = ["default"], workspace = true`. If the workspace enables default features though there's no way to disable default features.

Reply:
I see where my previous confusion came from now: I mentally swapped the first two columns of the table! Since I expected the package directive (the 'more local' one) to be higher precedence, I was intuitively expecting it further right in the table. I might also have not looked carefully enough at some rows.

Either way, I think it makes sense for the package directive to be able to override the workspace directive. That's just based on my thinking that that's the more general solution though. I don't think I've ever seen a workspace where most packages would want default features on, with only few turning them off for the same dependency.

Reply:
Ok I think I've addressed this with the most recent push, so resolving this.